### ISSUE TITLE
Random seed and reproducibility 

### ISSUE THREAD

**Issue by twangnh** [open-mmlab/mmdetection#2558]
mmdetection has an option for setting random seed:
```
def set_random_seed(seed, deterministic=False):
    """Set random seed.

    Args:
        seed (int): Seed to be used.
        deterministic (bool): Whether to set the deterministic option for
            CUDNN backend, i.e., set `torch.backends.cudnn.deterministic`
            to True and `torch.backends.cudnn.benchmark` to False.
            Default: False.
    """
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    if deterministic:
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False

```
But as noted in https://github.com/open-mmlab/mmdetection/issues/1698, the model training is not reproducible and resulting performance varies dramatically, I also tried to set the seed, but model training process is not consistent for same seed of different runs, can anyone suggest how to fix the possible bug in current mmdetection implementation to have reproducible training?

**Comment by nnop** [comment]
Have you set `deterministic=True`?

**Comment by hellock** [comment]
We tried multiple runs for the same seed and found the results stable (smaller gap than 0.2). If you specify the seed via the command line, the seed of each process and dataloader worker will be set, please also refer to [this](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/datasets/builder.py#L123).
`deterministic=False` does not guarantee the same output with the same input, while `deterministic=True` does but slows down the speed. Even `deterministic=True` is set, the backpropagation may also be non-deterministic due to some CUDA implementation, but the difference is small.

**Comment by mbhurtel** [comment]
If anyone is still wondering, we can set the randomness in the **config** itself by simply using the following line:
```py
randomness=dict(seed=860880352)  # check the seed as described below
```

Once you run a model, you can check the log where you can find the following:
```
System environment:
    sys.platform: linux
    Python: 3.8.20 (default, Oct  3 2024, 15:24:27) [GCC 11.2.0]
    CUDA available: True
    MUSA available: False
    numpy_random_seed: 860880352
    ...

Runtime environment:
    cudnn_benchmark: True
    mp_cfg: {'mp_start_method': 'fork', 'opencv_num_threads': 0}
    dist_cfg: {'backend': 'nccl'}
    seed: 860880352
    ...
```

Put the same seed when training again later. When you run the code with the same seed,  you will get exactly the same training results!

Note: The reproducibility will be seen if tested on the same GPU, but will differ if tested on a different GPU!_
